### PR TITLE
feat: Vec data structure

### LINKED ISSUE #55: feat: vec data structure
# Feature Request

**Describe the Feature Request**
Implement a dict-based vec data structure. Basically a mutable array.

**Describe Preferred Solution**

**Describe Alternatives**

**Related Code**

**Additional Context**

**If the feature request is approved, would you be willing to submit a PR?**
_(Help can be provided if you need assistance submitting a PR)_

- [ ] Yes
- [ ] No

### PR DESCRIPTION
**Pull request by enitrat**
## Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Issue Number: Closes #55 

## What is the new behavior?

- Mutable vec data structure
-
-

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: quaireaux/data_structures/src/vec.cairo [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+    /// Returns the item at the given index, or panics if the index is out of bounds.
+    /// Parameters
+    /// * self The vec instance.
+    /// * index The index of the item to get.
+    /// Returns
+    /// * T The item at the given index.
+    fn at(ref self: Vec<T>, index: usize) -> T {
+        if index < self.len() {
+            let item = self.items.get(index.into());
```
Comment (gaetbout) [inline quaireaux/data_structures/src/vec.cairo:75]: I'd directly return:
`self.items.get(index.into())`

File: quaireaux/data_structures/src/vec.cairo [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+
+trait VecTrait<T> {
+    fn new() -> Vec<T>;
+    fn get(ref self: Vec<T>, index: usize) -> Option<T>;
+    fn at(ref self: Vec<T>, index: usize) -> T;
+    fn push(ref self: Vec<T>, value: T) -> ();
+    fn set(ref self: Vec<T>, index: usize, value: T);
+    fn len(self: @Vec<T>) -> usize;
+}
```
Comment (wraitii) [inline quaireaux/data_structures/src/vec.cairo:41]: IMO you should define this as such:
```rust
ait VecTrait<U, T> {
     fn new() -> U;
     fn get(ref self: U, index: usize) -> Option<T>;
     fn at(ref self: U, index: usize) -> T;
     fn push(ref self: U, value: T) -> ();
     fn set(ref self: U, index: usize, value: T);
     fn len(self: @U) -> usize;
 }
```

**Comment by LucasLvy** [comment]
Fix la ci stp

**Comment by gaetbout** [comment]
> Fix la ci stp
He's on a more recent version of Cairo.
We'll need release 8 to be able to merge this

**Comment by enitrat** [comment]
> > Fix la ci stp
> 
> He's on a more recent version of Cairo. We'll need release 8 to be able to merge this

Haven't you seen? They released 1.0.0-rc0 a few days ago. Once quaireaux is updated to this version, I'll make the required changes :)

**Comment by gaetbout** [comment]
Is it considered as an "official release" I heard there should be a more official release next week maybe best to wait for that one?

**Comment by enitrat** [comment]
> Is it considered as an "official release"? I heard there should be a more official release next week maybe best to wait for that one?

release candidate are even more "official" than alpha versions, no?

**Comment by gaetbout** [comment]
I'll work on updating quaireaux to that one right away ser 🙏
Let me quickly check with other PR if it won't create issues

**Comment by enitrat** [comment]
@gaetbout ready to merge :)

**Review by abdelhamidbakhta — APPROVED** [review]
lgtm

File: quaireaux/data_structures/src/vec.cairo (hunk 2 above)
Comment (wraitii) [reply quaireaux/data_structures/src/vec.cairo:41]: J'avais pas vu ta réponse, sorry. Ba parce que le concept de ton interface 'vecteur' est pas propre au vecteur, mais juste à un conteneur à l'interface similaire. Si je veux une struct qui a la même interface mais pas le même layout en mémoire, je peux pas le faire.

**Pull request merged**